### PR TITLE
Dockerfile.deploy: use multistage docker build

### PR DESCRIPTION
**Pull request by vrutkovs**
Build `deploy` container using builder and artifacts stages to make it slim

### CODE REVIEW

**Comment by LalatenduMohanty** [comment]
Does podman supports multi-stage builds yet. I will be interested  build it with podman as it is more sustainable in long term for us.

**Comment by vrutkovs** [comment]
podman supports multistage builds - and so does imagebuilder used in OCP

File: dist/Dockerfile.build/Dockerfile [hunk 1]
```diff
@@ -32,3 +32,5 @@ RUN \
   mkdir -p $HOME/.cargo/git/ && \
   find $HOME/. -type d -exec chmod 777 {} \; && \
   find $HOME/. -type f -exec chmod ugo+rw {} \;
+
+WORKDIR /go/src/github.com/openshift/cincinnati
```
Comment (steveej) [inline dist/Dockerfile.build/Dockerfile]: Is this GOPATH structure as the CI is built mainly for Go packages, and it clones the source accordingly?

File: dist/Dockerfile.deploy/Dockerfile [hunk 2]
```diff
@@ -1,3 +1,11 @@
+FROM cincinnati:build as builder
+# build
+COPY . .
+RUN cargo build --release && \
+    mkdir -p /opt/cincinnati/bin && \
+    cp -rvf target/release/graph-builder /opt/cincinnati/bin && \
+    cp -rvf target/release/policy-engine /opt/cincinnati/bin && \
+
```
Comment (steveej) [inline dist/Dockerfile.deploy/Dockerfile:8]: Either
```suggestion
    :
```
or remove the `&& \` in the previous line

Comment (vrutkovs) [reply dist/Dockerfile.deploy/Dockerfile:8]: Yep, fixed, thanks

File: dist/Dockerfile.build/Dockerfile (hunk 1 above)
Comment (vrutkovs) [reply dist/Dockerfile.build/Dockerfile]: Yes, that's the path CI would put source at. I don't think its required though - removed

**Comment by steveej** [comment]
/lgtm
Not sure about the timing for merging this PR, so:
/hold

File: hack/e2e.sh [hunk 3]
```diff
@@ -23,12 +23,10 @@ oc create secret generic ci-pull-secret --from-file=.dockercfg=/tmp/cluster/pull
 oc secrets link default ci-pull-secret --for=pull
 
 # Apply oc template
-oc new-app -f dist/openshift/cincinnati.yaml \
+oc new-app -f ../template/cincinnati.yaml \
```
Comment (steveej) [inline hack/e2e.sh]: It's not clear to me why this path is relative to `hack/`

**Review by steveej — APPROVED** [review]
/lgtm
one question though:
/hold

File: hack/e2e.sh (hunk 3 above)
Comment (vrutkovs) [reply hack/e2e.sh]: This is set by `WORKDIR` in dockerfile:
```
WORKDIR "${HOME}/cincinnati"
...
COPY --from=builder hack/e2e.sh hack/
COPY --from=builder dist/openshift/cincinnati.yaml template/
```

Comment (vrutkovs) [reply hack/e2e.sh]: Ah, I was wrong - `hack/e2e.sh` would not change pwd to `hack`. Fixed

**Comment by vrutkovs** [comment]
/hold cancel

e2e test is now passing!

**Pull request merged**